### DISCUSSION TITLE
Optical Character Recognition (OCR)

### DISCUSSION THREAD

**Discussion by busterbeam** [streetcomplete/StreetComplete#4222] [Ideas]
Sometimes taking photos can look weird other times not so much, for example taking photos of opening times.  Standing by a complex opening time table can be quite the task to enter in.  Can OCR be applied to reading timetables?

I know time tables can very greatly from place to place but as long as it can speed things up.  You could have interactive boxes appear around each time suggesting what it means and from the first "correction" it could guess what the rest of the entries could likely be.

Just a thought

**Comment by 10992-osm** [comment]
What do you mean by "Sometimes taking photos can look weird"?  Does/would OCR somehow solve that?

**Comment by mnalis** [comment]
- You can choose `Other Answers` / `Can't say` / `Leave note` to take a picture of opening hours and leave it as a OSM note.
- For automatically OCR-ing it, see https://github.com/streetcomplete/StreetComplete/issues/1186

**Reply by busterbeam** [reply]
I know you can take pictures.  But it has been 2 years since that issue and I really do think is doable.  I do wish I had more kotlin experience but at the same time I wouldn't want either.

**Reply by mnalis** [reply]
@busterbeam have you read that issue again, especially comments near the end? 

There are reasons it is not done - it fails miserably even on many many simple cases, much less on complex ones.
Even if OCR part were to work absolutely flawlessly (which likely won't happen for years/decades after we've transited to fully autonomous cars) and you get text as was written (often handwritten!), trying to convert that text table to `opening_hours=*` format is very hard (you can try yourself online at https://webmapping.cyou/WebToOSMOH/ for example)

You say you think it is doable; have you found any site online or any other program which is able to do that with any kind of precision? Or do you have non-kotlin experience programming similar things? If you can show proof-of-concept in any technology (e.g. web CGI or whatever), it might renew interest in it.

**Reply by busterbeam** [reply]
So you're mo so pointing at the problem that even webpages that "describe" opening times in a non standardised way is "impossible" let alone OCR.

I guess it's a matter of a statistical model "trained" on text.  We need the two "neural" nets, one which gets text from image and one that fills the field data from the text.  And make the model nice and small to fit in the app nice and tight

**Reply by busterbeam** [reply]
Anywhere can I get a bunch of images of opening times signs for training data

**Reply by mnalis** [reply]
> So you're mo so pointing at the problem that even webpages that "describe" opening times in a non standardised way is "impossible" let alone OCR.

Yes. It's far from impossible - human neural net seems to be quite capable of dealing with it. So it is "only" a matter of advanced enough technology/methods, sufficient resources (either locally on phone, or connecting to some "cloud" service), gathering enough samples, and training / validating data until it is **at least** 95%+ accurate on random sample. 

So at least two different neural nets are needed:

- one that can correctly OCR (handling images taken from different angles, different lightning, reflections, shadows, colors, fonts, handwriting, etc.) to produce near perfect text from image
- another that can interpret such text formatted for humans, and then convert it to opening_hours format.   

> Anywhere can I get a bunch of images of opening times signs for training data

sure, for example:
- https://duckduckgo.com/?q=opening+hours+sign&hps=1&iax=images&ia=images
- search https://ent8r.github.io/NotesReview/ for `opening hours` and find image links (e.g. https://westnordost.de/p/86304.jpg)

Give it a try by all means! I do not mean to discourage it, but I'd be superbly surprised if you can get anywhere near usable results on either of those AI neural nets, much less their coordination. (and If you do, I've heard that AI FullSelfDrive lead position at Tesla has recently been vacated, and I'm sure they'd love to have you onboard!)
